group_theory/perm.lean

```diff
@@ -286,9 +385,204 @@ funext $ λ f,
+  λ g' hg',
+  let ⟨g, hg⟩ := list.mem_map.1 hg' in
+  hg.2 ▸ is_swap_of_subtype (l.2.2 _ hg.1),
+have hl'₂ : (l.1.map of_subtype).prod = f,
+  by rw [← is_group_hom.prod of_subtype l.1, l.2.1, of_subtype_subtype_perm _ h₂],
+by rw [sign_eq_pow_length_swap_factors l.2.1 l.2.2, sign_eq_pow_length_swap_factors hl'₂ hl'];
+  simp
+
+@[simp] lemma sign_of_subtype [fintype α] {p : α → Prop} [decidable_pred p]
+  (f : perm (subtype p)) : sign (of_subtype f) = sign f :=
+have ∀ x, of_subtype f x ≠ x → p x, from λ x, not_imp_comm.1 (of_subtype_apply_of_not_mem f),
+by conv {to_rhs, rw [← subtype_perm_of_subtype f, sign_subtype_perm _ _ this]}
+
+lemma sign_eq_sign_of_equiv [fintype α] [decidable_eq β] [fintype β] (f : perm α) (g : perm β)
+  (e : α ≃ β) (h : ∀ x, e (f x) = g (e x)) : sign f = sign g :=
+have hg : g = (e.symm.trans f).trans e, from equiv.ext _ _ $ by simp [h],
+by rw [hg, sign_symm_trans_trans]
+
+lemma sign_bij [fintype α] [decidable_eq β] [fintype β]
+  (f : perm α) (g : perm β) (i : Π x : α, f x ≠ x → β)
+  (h : ∀ x hx hx', i (f x) hx' = g (i x hx))
+  (hi : ∀ x₁ x₂ hx₁ hx₂, i x₁ hx₁ = i x₂ hx₂ → x₁ = x₂)
+  (hg : ∀ y, g y ≠ y → ∃ x hx, i x hx = y) :
+  sign f = sign g :=
+calc sign f = sign (@subtype_perm _ f (λ x, f x ≠ x) (by simp)) :
+eq.symm (sign_subtype_perm _ _ (λ _, id))
+... = sign (@subtype_perm _ g (λ x, g x ≠ x) (by simp)) :
+sign_eq_sign_of_equiv _ _
+  (equiv.of_bijective
+    (show function.bijective (λ x : {x // f x ≠ x},
+      (⟨i x.1 x.2, have f (f x) ≠ f x, from mt (λ h, f.bijective.1 h) x.2,
+        by rw [← h _ x.2 this]; exact mt (hi _ _ this x.2) x.2⟩ : {y // g y ≠ y})),
+      from ⟨λ ⟨x, hx⟩ ⟨y, hy⟩ h, subtype.eq (hi _ _ _ _ (subtype.mk.inj h)),
+        λ ⟨y, hy⟩, let ⟨x, hfx, hx⟩ := hg y hy in ⟨⟨x, hfx⟩, subtype.eq hx⟩⟩))
+    (λ ⟨x, _⟩, subtype.eq (h x _ _))
+... = sign g : sign_subtype_perm _ _ (λ _, id)
+
+lemma is_cycle_swap {x y : α} (hxy : x ≠ y) : is_cycle (swap x y) :=
+⟨y, by rwa swap_apply_right,
+λ a (ha : ite (a = x) y (ite (a = y) x a) ≠ a),
+if hya : y = a then ⟨0, hya⟩
+  else ⟨1, by rw [gpow_one, swap_apply_def]; split_ifs at *; cc⟩⟩
+
+lemma is_cycle_inv {f : perm α} (hf : is_cycle f) : is_cycle (f⁻¹) :=
+let ⟨x, hx⟩ := hf in
+⟨x, by simp [eq_inv_iff_eq, inv_eq_iff_eq, *] at *; cc,
+  λ y hy, let ⟨i, hi⟩ := hx.2 y (by simp [eq_inv_iff_eq, inv_eq_iff_eq, *] at *; cc) in
+  ⟨-i, by rwa [gpow_neg, inv_gpow, inv_inv]⟩⟩
+
+lemma swap_mul_eq_mul_swap (f : perm α) (x y : α) : swap x y * f = f * swap (f⁻¹ x) (f⁻¹ y) :=
+equiv.ext _ _ $ λ z, begin
+  simp [mul_apply, swap_apply_def],
+  split_ifs;
+  simp [*, eq_inv_iff_eq] at * <|> cc
+end
+
+lemma mul_swap_eq_swap_mul (f : perm α) (x y : α) : f * swap x y = swap (f x) (f y) * f :=
+by rw [swap_mul_eq_mul_swap, inv_apply_self, inv_apply_self]
+
+lemma is_cycle_swap_mul_aux : ∀ (n : ℤ) {b x : α} {f : perm α} (hf : is_cycle f)
+  (hb : (swap x (f x) * f) b ≠ b)
+  (h : (f ^ n) (f x) = b), ∃ i : ℤ, ((swap x (f x) * f) ^ i) (f x) = b
```

Comment:
Write `hf`, `hb`, and `h` in one line before the actual goal. You may also replace `hf`, `hb`, and `h` by implications as the names cannot be used.